resharper/resharper-unity/src/Rider/CodeInsights/RiderUnityImplicitUsageHighlightingContributor.cs

```diff
@@ -0,0 +1,72 @@
+    {
+        private readonly UnityImplicitFieldUsageProvider myFieldUsageProvider;
+        private readonly UnityImplicitCodeInsightProvider myImplicitCodeInsightProvider;
+        private readonly IconHost myIconHost;
+
+        public RiderUnityImplicitUsageHighlightingContributor(ISolution solution, ITextControlManager textControlManager,
+            UnityImplicitFieldUsageProvider fieldUsageProvider,  UnityImplicitCodeInsightProvider implicitCodeInsightProvider,
+            ISettingsStore settingsStore, IconHost iconHost = null)
+            : base(solution, settingsStore, textControlManager)
+        {
+            myFieldUsageProvider = fieldUsageProvider;
+            myImplicitCodeInsightProvider = implicitCodeInsightProvider;
+            myIconHost = iconHost;
+        }
+        
+        public override void AddHighlighting(IHighlightingConsumer consumer, ICSharpDeclaration element, string tooltip)
+        {
+            switch (element)
+            {
+                case IFieldDeclaration _:
+                    AddHighlighting(consumer, myFieldUsageProvider, element, tooltip, "Set by Unity");
+                    break;
+                case IClassLikeDeclaration _:
+                    AddHighlighting(consumer, myImplicitCodeInsightProvider, element, tooltip, "Scripting component");
+                    break;
+                case IMethodDeclaration _:
+                    AddHighlighting(consumer, myImplicitCodeInsightProvider, element, tooltip, "Event function");
+                    break;
+                default:
+                    AddHighlighting(consumer, myImplicitCodeInsightProvider, element, tooltip, "Implicit usage");
+                    break;
+            }
+        }
+        
+        private void AddHighlighting(IHighlightingConsumer consumer, AbstractUnityImplicitProvider codeInsightsProvider, ICSharpDeclaration element, string tooltip, string displayName)
+        {
+            if (SettingsStore.GetIndexedValue((CodeInsightsSettings key) => key.DisabledProviders, codeInsightsProvider.ProviderId))
+            {
+                base.AddHighlighting(consumer, element, tooltip);
+                return;
+            }
+
+            if (SettingsStore.GetValue((UnitySettings key) => key.GutterIconMode) == GutterIconMode.Always)
+            {
+                base.AddHighlighting(consumer, element, tooltip);
+            }
+            
+            displayName = displayName ?? codeInsightsProvider.DisplayName;
+            var declaredElement = element.DeclaredElement;
+            if (declaredElement == null || !declaredElement.IsValid())
+                return;
+            
+            consumer.AddHighlighting(new UnityCodeInsightsHighlighting(element.NameIdentifier.GetDocumentRange(),
+                displayName, displayName, codeInsightsProvider, declaredElement, myIconHost.Transform(codeInsightsProvider.IconId), CreateBulbItemsForUnityDeclaration(element)));
```

Comment:
Do we need to call this `Transform` method? I haven't seen it used in other places we use icons.

Reply:
Passing through `displayName` twice. Should one of those be `tooltip`?